lib/rust/parser/cdylib/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+/// `parser_free`.
+/// The `input` argument must be a non-NULL pointer to a sequence of at least `input_len` readable
+/// bytes.
+/// The `output` argument must be a value returned by `parse_result_alloc` that has never been
+/// passed to `parse_result_free`.
+#[no_mangle]
+pub unsafe extern "C" fn parser_run(
+    parser: *mut Parser,
+    input: *const u8,
+    input_len: usize,
+    output: *mut ParseResult,
+) {
+    let parser = &mut *parser;
+    let input = slice::from_raw_parts(input, input_len);
+    let output = &mut *output;
+    *output = parser.run(input);
+}
+
+/// Release the parser's resources.
+///
+/// # Safety
+///
+/// The `parser` argument must be a value returned by `parser_new` that has never been passed to
+/// `parser_free`.
+#[no_mangle]
+pub unsafe extern "C" fn parser_free(parser: *mut Parser) {
+    let parser = Box::from_raw(parser);
+    drop(parser);
+}
+
+
+
+// ===================
+// === ParseResult ===
+// ===================
+
+/// The result of parsing. Contains the source code, if the parser modified it, and the serialized
+/// AST.
+#[derive(Debug, Default)]
+pub struct ParseResult {
+    input: Option<String>,
+    tree:  Vec<u8>,
+}
+
+
+// === C Interface ===
+
+/// Allocate a new parse result; its initial state is unspecified. This operation is infallible; the
+/// result will never be NULL.
+#[no_mangle]
+pub extern "C" fn parse_result_alloc() -> *mut ParseResult {
+    Box::into_raw(default())
+}
+
+/// If the input provided was valid utf-8, this will return NULL. If it was not, this will return
+/// the beginning of a copy of the input that has been modified to contain only valid utf-8. If this
+/// returns non-NULL, all source code in the serialized AST data will be relative to this pointer,
+/// not the original input.
+///
+/// # Safety
+///
+/// The `result` argument must be a value returned by `parse_result_alloc` that has never been
+/// passed to `parse_result_free`.
+#[no_mangle]
+pub unsafe extern "C" fn parse_result_get_input_data(result: *mut ParseResult) -> *const u8 {
+    let result = &*result;
+    result.input.as_ref().map(|s| s.as_ptr()).unwrap_or(0 as _)
+}
+
+/// Return the length of the data obtained by [`parse_result_get_input_data`].
+///
+/// # Safety
+///
+/// The `result` argument must be a value returned by `parse_result_alloc` that has never been
+/// passed to `parse_result_free`.
+#[no_mangle]
+pub unsafe extern "C" fn parse_result_get_input_len(result: *mut ParseResult) -> usize {
+    let result = &*result;
+    result.input.as_ref().map(|s| s.len()).unwrap_or(0)
+}
+
+/// Return a pointer to the start of the serialized AST data.
+///
+/// # Safety
+///
+/// The `result` argument must be a value returned by `parse_result_alloc` that has never been
+/// passed to `parse_result_free`.
+#[no_mangle]
+pub unsafe extern "C" fn parse_result_get_output_data(result: *mut ParseResult) -> *const u8 {
```

Comment:
What is the order of this array? Is it `LITTLE_ENDIAN` or `BIG_ENDIAN` or platform dependent?

Reply:
It is always LE.